labelbox/data/annotation_types/base_annotation.py

```diff
@@ -1,10 +1,18 @@
 import abc
+from uuid import UUID, uuid4
 from typing import Any, Dict
+from pydantic import PrivateAttr
 
 from .feature import FeatureSchema
 
 
 class BaseAnnotation(FeatureSchema, abc.ABC):
     """ Base annotation class. Shouldn't be directly instantiated
     """
+    _uuid: UUID = PrivateAttr()
     extra: Dict[str, Any] = {}
+
+    def __init__(self, **data):
+        super().__init__(**data)
+        extra_uuid = data.get("extra", {}).get("uuid")
+        self._uuid = data.get("_uuid") or extra_uuid or uuid4()
```

Comment:
good question! so, before the UUIDs were generated only during serialization (python annotation -> NDJSON), which could have worked for relationships but would have made it more complicated to figure out the UUIDs at that point. First, I'd need to keep track of the resulting NDJSON object for each of the annotations involved in the relationships, keeping a dict to store annotation -> NDJSON and then I'd be able to access the UUID of the resulting NDJSON once I serialize the relationship annotations. The idea was to allow the user to create a relationship annotation object by using references to the related objects, like this: instead of  The reason was so we

Reply:
good question! so, before the UUIDs were generated only during serialization (python annotation -> NDJSON), which could have worked for relationships but would have made it more complicated to figure out the UUIDs at that point. First, I'd need to keep track of the resulting NDJSON object for each of the annotations involved in the relationships, keeping a dict to store annotation -> NDJSON and then I'd be able to access the UUID of the resulting NDJSON once I serialize the relationship annotations. The idea was to allow the user to create a relationship annotation object by using references to the related objects, like this: instead of  The reason was so we

Reply:
Good question!

So, before we generated the UUIDs only during serialization (python annotation -> NDJSON), which could have worked for relationships but would have made it more complicated to figure out the UUIDs at that point. First, I'd need to keep track of the resulting NDJSON object for each of the annotations involved in the relationships, keeping a dict to store annotation -> NDJSON and then I'd be able to obtain the UUIDs of the resulting NDJSON objects during the serialization of the relationship annotations. As serialization is heavily used everywhere (more than deserialization, which probably users don't use or at least not as much), it was necessary to keep things smoother on the serialization side and it was a lot easier to get the UUIDs of the related objects if that UUID was already in the annotation object itself without the need to keep additional state of the already serialized objects (we just `yield` those NDJSON objects as soon as they are ready, would have taken more memory if some of that information needed to stay around just for relationships).

Keep in mind that the idea was to keep things simple on the user's side and allow them to create a relationship annotation object by using references to the related objects, like this: 
```
source_object = {...}
target_object = {...}
relationship = {
   source: source_object,
   target: target_object
}
```
instead of making them generate UUIDs on their side and passing them as strings in the relationship objects:
```
from uuid import uuid4
source_object = {uuid: "27c55635-a5c3-4826-9436-7e95046718d5" or uuid4(), ...}
target_object = {uuid: "b4ca3d52-340d-4d3f-bbe3-a026ba1818c5" or uuid4(), ...}
relationship = {
   source: source_object.uuid,
   target: target_object.uuid
}
```
Additionally, annotation objects didn't have an `uuid` field before just an `extra["uuid"]` and instead of adding a new public `uuid` field (which would have probably been confusing to them b/c of the other extra.uuid field) so they can add those UUIDs themselves, I decided to add this field and its implementation/usage hidden from them and also hidden from any serialization's output.